### PR TITLE
Support /mnt/ and \\wsl.localhost\ and \\wsl$\ style links on Windows frontends

### LINKED ISSUE #148062: Terminal links should support WSL folders
This text in terminal should be clickable and open Windows explorer with WSL path:
```
//wsl$/Ubuntu/tmp/tmpxjck8ple
```

Currently it isn't clickable.

Similar issue to: https://github.com/microsoft/vscode/issues/102638

### PR DESCRIPTION
**Pull request by Tyriar**
Fixes https://github.com/microsoft/vscode/issues/148062

image
image

### CODE REVIEW

**Comment by Tyriar** [comment]
@aeschli a review would be appreciated since you're fairly familiar with WSL

**Comment by aeschli** [comment]
Here's the code I use in the WSL extension. 
- It has some more backward compatibility things checking for older WSL verisons, but after all these years I think we can skip these
- It avoids backsplash escaping issues by using env variables to make the conversion

```
export function executeWslPath(pathToConvert: string, toLinux: boolean, distro: string): string {
	if (toLinux) {
		let commandLine = 'echo "$VSCODE_WSL_PATH"';
		const options = {
			'WSLENV': 'VSCODE_WSL_PATH/up',
			'VSCODE_WSL_PATH': pathToConvert
		};
		return executeCommand(commandLine, options, distro).trim();
	} else {
		let commandLine = `wslpath -w "$VSCODE_WSL_PATH"`;
		const options = {
			'WSLENV': 'VSCODE_WSL_PATH/u',
			'VSCODE_WSL_PATH': pathToConvert
		};
		return executeCommand(commandLine, options, distro).trim();
	}
}

export function executeCommand(commandLine: string, env: Record<string, string>, distro: string): string {
	const wslExecPath = getWSLExecutablePath();
	if (!wslExecPath) {
		throw new Error(localize('executeCommand.noWSL', 'wsl command not found\n\nMake sure WSL is installed:\nhttps://aka.ms/vscode-remote/wsl/install-wsl'));
	}
	const options = getExecSyncOptions(env);
	let args = [];
	if (isExecAndDistroArgumentSupported()) {
		args.push('-d', distro, '-e', 'sh', '-c', commandLine);
	} else if (path.basename(wslExecPath).toLowerCase() === 'wsl.exe') {
		args.push('sh', '-c', commandLine);
	} else {
		args.push('-c', commandLine);
	}
	return cp.execFileSync(wslExecPath, args, options as cp.ExecFileSyncOptions).toString();
}

export function getWSLExecutablePath(): string | undefined {
	let useWSLexe = getWindowsBuildNumber() >= 16299;
	const is32ProcessOn64Windows = process.env.hasOwnProperty('PROCESSOR_ARCHITEW6432');
	const systemRoot = process.env['SystemRoot'];
	if (systemRoot) {
		return path.join(systemRoot, is32ProcessOn64Windows ? 'Sysnative' : 'System32', useWSLexe ? 'wsl.exe' : 'bash.exe');
	}
	return undefined;
}

export function isExecAndDistroArgumentSupported(): boolean {
	// https://docs.microsoft.com/en-us/windows/wsl/release-notes#build-17666
	return getWindowsBuildNumber() >= 17666; // part of September 2018 Update (1809)
}

export function getExecSyncOptions(env: Record<string, string> = {}): cp.ExecSyncOptions & { env: NodeJS.ProcessEnv; stdio: cp.StdioOptions } {
	return { env, stdio: [null, null, null] };
}
```

**Review by aeschli — CHANGES_REQUESTED** [review]

**Comment by aeschli** [comment]
Your changes look ok, I just put the `request changes` so you can review my comments.
Looks like `escapeNonWindowsPath` bans some characters that are not forbidden in paths. In lInux the only character that is forbidden is `/`. On Windows it's more but not  `$`. That could be a problem with `\\wsl$\...`
 I'd suggest to use the same code I use in the extension.

File: src/vs/workbench/contrib/terminal/browser/links/terminalLinkManager.ts [hunk 1]
```diff
@@ -463,16 +463,30 @@ export class TerminalLinkManager extends DisposableStore {
...
 			return null;
 		}
 
-		const linkUrl = removeLinkSuffix(preprocessedLink);
-		if (!linkUrl) {
-			this._resolvedLinkCache.set(link, null);
-			return null;
+		// If the link looks like a /mnt/ WSL path and this is a Windows frontend, use the backend
+		// to get the resolved path from the wslpath util.
+		if (isWindows && link.match(/^\/mnt\/[a-z]/i) && this._processManager.backend) {
```
[the remark this answers is not in the document]
Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/links/terminalLinkManager.ts:475]: I think this sort of thing would have to be a separate feature request. It should just work out of the box in a remote wsl workspace or if you have the local wsl folder open, but to reliably know what distro to do we would want to have more knowledge about what's actually running in the terminal. That's possible but out of the scope of this change imo.

**Comment by Tyriar** [comment]
`escapeNonWindowsPath` is an aggressive function for preventing script injection across many shells, sure it doesn't work with some paths but it's what the terminal uses to sanitize paths for security.

**Comment by Tyriar** [comment]
Although we may not need to worry about injection when using execFile vs exec 🤔

**Comment by aeschli** [comment]
execFile is safer. Alos it's important to run it with the full path. Just invoking `bash.exe` can be a problem.

File: src/vs/workbench/contrib/terminal/browser/links/terminalLinkManager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (aeschli) [reply src/vs/workbench/contrib/terminal/browser/links/terminalLinkManager.ts:475]: ok, understand that you don't know the distro, especially in a local window, when the user uses `wsl.exe' himself. And just looking for `/mnt/` and calling wsl.exe of the default distro will probably work, but just to say, distros can change the windows root from '/mnt/` to [something else](https://learn.microsoft.com/en-us/windows/wsl/wsl-config#automount-settings)
In a remote WSL window all that should not really be necessary. The link can be detected as a linux path `/mnt/...` and opened as such (ill open and editor on '/mnt/...`)

**Review by aeschli — APPROVED** [review]

**Pull request merged**